File: bindings/java/src/test/java/org/wickra/screener/ScreenerTest.java
package org.wickra.screener;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class ScreenerTest {
    private static final String SPEC =
            "{\"universe\":[\"AAA\",\"BBB\"],\"condition\":{\"type\":\"cmp\","
                    + "\"left\":{\"kind\":\"price\",\"field\":\"close\"},\"op\":\"gt\","
                    + "\"right\":{\"kind\":\"const\",\"value\":10.0}}}";

    private static String candle(int close) {
        return "{\"time\":1,\"open\":" + close + ",\"high\":" + close
                + ",\"low\":" + close + ",\"close\":" + close + ",\"volume\":1}";
    }

    @Test
    void versionIsNonEmpty() {
        assertFalse(Screener.version().isEmpty());
    }

    @Test
    void scanReturnsMatchingSymbol() {
        try (Screener screener = new Screener(SPEC)) {
            String cmd = "{\"cmd\":\"scan\",\"data\":{"
                    + "\"AAA\":[" + candle(5) + "],"
                    + "\"BBB\":[" + candle(15) + "]}}";
            String raw = screener.command(cmd);
            assertTrue(raw.contains("\"scanned\":2"), raw);
            assertTrue(raw.contains("\"symbol\":\"BBB\""), raw);
            assertFalse(raw.contains("\"symbol\":\"AAA\""), raw);
        }
    }

    @Test
    void invalidSpecThrows() {
        assertThrows(IllegalArgumentException.class, () -> new Screener("not json"));
    }

    @Test
    void unknownCommandIsInBandError() {
        try (Screener screener = new Screener(SPEC)) {
            // An unknown command is not a hard error: the ABI returns a length and
            // the error surfaces in-band as {"ok":false,...} JSON.
            String raw = screener.command("{\"cmd\":\"nope\"}");
            assertEquals(true, raw.contains("\"ok\":false"), raw);
        }
    }
}

File: bindings/node/__tests__/screener.test.js
"use strict";

const { test } = require("node:test");
const assert = require("node:assert");
const { Screener, version } = require("../index.js");

const SPEC = JSON.stringify({
  universe: ["AAA", "BBB"],
  condition: {
    type: "cmp",
    left: { kind: "price", field: "close" },
    op: "gt",
    right: { kind: "const", value: 10.0 },
  },
});

function candle(close) {
  return { time: 1, open: close, high: close, low: close, close, volume: 1.0 };
}

test("scan roundtrip returns only the matching symbol", () => {
  const screener = new Screener(SPEC);
  const response = screener.command(
    JSON.stringify({
      cmd: "scan",
      data: { AAA: [candle(5.0)], BBB: [candle(15.0)] },
    }),
  );
  const report = JSON.parse(response);
  assert.strictEqual(report.scanned, 2);
  assert.deepStrictEqual(
    report.matches.map((m) => m.symbol),
    ["BBB"],
  );
});

test("version matches the module-level function", () => {
  const screener = new Screener(SPEC);
  assert.strictEqual(screener.version(), version());
});

test("a malformed spec throws", () => {
  assert.throws(() => new Screener("not json"));
});
